src/java/org/apache/cassandra/index/sai/utils/IndexFileUtils.java

```diff
@@ -101,6 +102,13 @@ class IncrementalChecksumSequentialWriter extends SequentialWriter implements Ch
             super(file, writerOption);
         }
 
+        @Override
+        public void write(ByteBuffer src) throws IOException
+        {
+            super.write(src.slice());
```

Comment:
I think the first copy is unnecessary since the superclass implementation makes sure to not modify its parameter

Reply:
slice() doesn't copy the content. https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#slice()
It's a light-weight object that shares the original ByteBuffer's content but has independent position/limit/mark.
It guarantees that original buffer's position won't change.

Reply:
ByteBuffers are not really lightweight.

On the one hand it probably doesn't matter for the current use case where only large-ish buffers are involved and the BB [shallow] copying is relatively insignificant.  OTOH this seems like the kind of subtle difference that takes a few minutes to address now but hours to troubleshoot later.

Reply:
... I'm okay with leaving this for Future DataStax to worry about, if necessary.

But I do think we should remove the first copy.

Reply:
ok with removing first slice().
I still don't get the idea about "adding a writeAndConsume". checksum.update() is jdk CRC32's method, I don't see benefits of reimplementing its internals. 
"ByteBuffers are not really lightweight". ByteBuffs might not be lightweight but slice() does not copy the content.

I am probably missing some cassandra-specific concern.

Reply:
I'm just saying that .duplicate, .slice, .asFloatBuffer, and similar methods should not be called in a hot loop because the BB itself is a fairly heavy object even though contents are not copied.

If we had a writeAndConsume method then we wouldn't need the 2nd slice either because our contract would be "writes the buffer and exhausts it" instead of "writes the buffer without changing it."